docs/architecture/adr-007-light-client-supervisor-ergonomics.md

```diff
@@ -0,0 +1,53 @@
+# ADR 007: Light Client Supervisor Ergonomics
+
+## Changelog
+
+* 2020-06-30: Initial draft
+
+## Context
+
+The initial approach to use callbacks in order to capture async responses
+from requests issued by the `Handle` to its `Supervisor` introduces a couple of
+drawbacks, i.e. try semantics, scope and borrow complexities. As the
+general pattern seems to be that a bounded channel is constructed to transport
+the response it could be reworked to be the actual value send with the event to
+the Supervisor. Which in turn send back the response on the typed channel to the
+Handle. As an added side-effect this could remove the need for the Callback
+abstraction all together and fully embraces CSP style concurrency.
+
+Furthermore we have async versions of most `Handle` methods, but no real
```

Comment:
As mentioned in #403, The purpose of the async interface would be to facilitate the relayer dispatching verification from an event loop to allow events from queries/subscripts to accumulate concurrently. The result of the callback would then be written to the relayer event loop when verification results are ready.

Maybe this could also be done using async/await (as @romac  has suggested) however my understanding is that await calls within the relayer would block the relayer loop and defeat the concurrency that we are looking for. This discussion is probably outside the scope of this ADR but are there similarly simple (channels) methods of achieving the concurrency we want (accumulate events while waiting for verification) without implementing a scheduler?

Reply:
Interesting, I'd be curious to get more context on the claim of the blocking of the relayer loop.

Generally this sounds like the relayer will have very specific requirements for its concurrency, which in turn means driving the design of the interface of the `Handle` with it will likely result in brittle coupling. What should give us ease of mind is that that by design the `Handle` can be safely shared between threads, so whatever the relayer needs w.r.t. scheduling can be done on top of the exisiting synchronous methods. This will keep the surface small and guarantees flexibility on the caller side. We can codify and enforce that with marker traits (`Send` + `Sync`) and the already in use [static_assertions](https://docs.rs/static_assertions/1.1.0/static_assertions/).

Reply:
I think in my head, sequences of `await` calls are serialized even if there is no dependence:
```rust
loop {
  foo.compute().await;
  baz.compute().await;
}
```

Is this not the case? 

What we want is loop over a queue of unordered responses:

(excuse the pseudo-ish code here)
```rust
let sender, receiver = ...
while let Some(event) = receiver.recv() {
  foo.compute(sender.clone());
  baz.compute(sender.clone());
}
```

There could be a better way here, but this is the requirement we anticipate for the relayer.

Reply:
The above is related to the current concurrency architecture proposed in https://github.com/informalsystems/ibc-rs/issues/121

![image](https://user-images.githubusercontent.com/246/86227299-1f695100-bb8d-11ea-96bd-9c5a2f72503f.png)

Reply:
Thanks for the detailed explanations and extended context for the needs of downstream consumers. As I suspected the concurrency model of the relayer is quite specific and in essence would necessitate somethig like a [`select`](https://docs.rs/crossbeam-channel/0.4.2/crossbeam_channel/macro.select.html#examples) to provide the response to whatever computation has been completed. And again one can build that easily in the context of the ibc/relayer codebase with the synchronous API methods of the `Handle`. Esp. as you want to control the construction and type of the channels used to select over.

> I think in my head, sequences of `await` calls are serialized

That is correct, such is the nature of `await`.

Reply:
>  And again one can build that easily in the context of the ibc/relayer codebase with the synchronous API methods of the Handle

I believe the concurrency model we are looking for is one in which calls to the `chain light client` (and all other arrows) would specifically not block the `event handler`.

Reply:
I understnad the original design intend. My point is that the `_async` variants only partially alleviate the problem and introduce coupling with downstream consumers for not much gain. The non-blocking nature in the event loop of the relayer can be easily achieved in the context of the ibc repository on top of the synchronous methods, while leaving fine-grained control over the concurrency model to the implementation of the relayer. As an exmaple:

``` rust
let (first_sender, first_receiver) = channel::bounded::<_>(1);
let (second_sender, second_receiver) = channel::bounded::<_>(1);

thread::spawn(move || {
    let res = blocking_computation();
    first_sender.send(res);
});

thread::spawn(move || {
    let res = another_blocking_computation();
    second_sender.send(res);
});

loop {
    select!{
        recv(first_receiver) -> res => todo!(),
        recv(first_receiver) -> res => todo!(),
    }
}
```

Reply:
Makes sense. However this doesn't appear to embrace `async/await` or provide any clarity around the suggestion in the paragraph. Perhaps you could add a more illustrative example to the ADR or remove that particular comment if it's not critical to the major thrust of the argument.

Reply:
Apparently, you can also `select` over async functions / futures: https://rust-lang.github.io/async-book/06_multiple_futures/03_select.html